lightning-invoice/src/payment.rs

```diff
@@ -1364,24 +1532,290 @@ mod tests {
 		invoice_payer.handle_event(&event);
 	}
 
+	#[test]
+	fn generates_correct_inflight_map_data() {
+		let event_handled = core::cell::RefCell::new(false);
+		let event_handler = |_: &_| { *event_handled.borrow_mut() = true; };
+
+		let payment_preimage = PaymentPreimage([1; 32]);
+		let invoice = invoice(payment_preimage);
+		let payment_hash = Some(PaymentHash(invoice.payment_hash().clone().into_inner()));
+		let final_value_msat = invoice.amount_milli_satoshis().unwrap();
+
+		let payer = TestPayer::new().expect_send(Amount::ForInvoice(final_value_msat));
+		let final_value_msat = invoice.amount_milli_satoshis().unwrap();
+		let route = TestRouter::route_for_value(final_value_msat);
+		let router = TestRouter {};
+		let scorer = RefCell::new(TestScorer::new());
```

Comment:
Are these RefCells really needed? Wouldn't just creating TestScorers work generally?

Reply:
I believe this is because `InvoicePayer` expects the scorer to implement `LockableScore`. 

`TestScorer` here is really just a mocked version of a scorer that we'd expect users to implement on their own

Reply:
Yeah, a scorer needs to be locked to (a) use its mutable interface in `InvoicePayer`'s event handling and (b) use its immutable interface when routing such that its state doesn't change between calls. Additionally, another reference to it may be needed for serialization.

Reply:
I see, thanks! MinCostFlowRouter pull request is at a point where it needs to be tested together with InvoicePayer and tracking in-flight HTLCs, but in a conversion with Matt Corallo, he said that as MinCostFlowRouter needs to access liquidity tracking, but not a custom scorer provided by LDK users, LiquidityTracker should be a separate trait implemented by the scorers. Another option in my opinion would be keeping liquidity tracking from scoring totally separate, and DefaultRouter should request the estimated liquidity information and provide it to the immutable Scorer (that can be customized by LDK user, while liquidity tracking implementation probably shouldn't be surfaced in the API).

What do you guys think?

Here's a link to the comments:

https://github.com/lightningdevkit/rust-lightning/pull/1668#issuecomment-1230544540

Reply:
Yeah, I think we want at least two follow-ups to this PR:

(1) Move the `LockableScore` type parameter to `DefaultRouter` since some routers (e.g., remote server routing) don't need a scorer.
(2) Move liquidity tracking to either the `Payer` trait or a dedicated one, having `InvoicePayer` query it and pass the liquidity information to `Router` akin to how it currently passes a scorer.

The latter comes from an observation that I made in review club. `ChannelManager` actually knows all the  liquidity information given it is the one producing `PaymentPathFailed` and `PaymentPathSuccessful` events. It also would make liquidity tracking persist across restarts.

See also #1157 for exposing pending payments more generally, which should be considered when making this change.